src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -419,10 +423,38 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
 
             if (!type.IsValueType)
             {
-                bool hasDefaultConstructor = type.GetConstructors(DeclaredOnlyLookup).Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
-                if (!hasDefaultConstructor)
+                ConstructorInfo[] constructors = type.GetConstructors(DeclaredOnlyLookup);
+
+                bool hasParameterlessPublicConstructor = constructors.Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
+
+                if (!hasParameterlessPublicConstructor)
                 {
-                    throw new InvalidOperationException(SR.Format(SR.Error_MissingParameterlessConstructor, type));
+                    // if the only constructor is private, then throw
+                    if (constructors.Length == 1 && !constructors[0].IsPublic)
+                    {
+                        throw new InvalidOperationException(SR.Format(SR.Error_MissingParameterlessConstructor, type));
+                    }
+
+                    // find the biggest constructor so that we can bind to the most parameters
+                    ParameterInfo[] parameters = constructors[0].GetParameters();
+
+                    for (int index = 1; index < constructors.Length; index++)
+                    {
+                        ParameterInfo[] constructorParameters = constructors[index].GetParameters();
+                        if (constructorParameters.Length > parameters.Length)
+                        {
+                            parameters = constructorParameters;
+                        }
+                    }
+
+                    List<object?> parameterValues = new List<object?>();
+
+                    foreach (var parameter in parameters)
+                    {
+                        parameterValues.Add(GetParameterValue(parameter, config, options));
+                    }
+
+                    return Activator.CreateInstance(type, parameterValues.ToArray());
```

Comment:
We already have the constructor info here. Why not just call it? Activator would need to look it up again as it is written.

Reply:
Very good point! 👍